File: dqops/src/main/java/com/dqops/rest/models/metadata/RuleModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.rest.models.metadata;

import com.dqops.core.filesystem.virtual.FileContent;
import com.dqops.execution.rules.runners.python.PythonRuleRunner;
import com.dqops.metadata.definitions.rules.RuleDefinitionSpec;
import com.dqops.metadata.definitions.rules.RuleDefinitionWrapper;
import com.dqops.metadata.definitions.rules.RuleRunnerType;
import com.dqops.metadata.definitions.rules.RuleTimeWindowMode;
import com.dqops.metadata.fields.ParameterDefinitionsListSpec;
import com.dqops.rules.RuleTimeWindowSettingsSpec;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.Map;
import java.util.Objects;
import java.util.Optional;

/**
 * Rule model that is returned by the REST API. Describes a single unique rule name.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "RuleModel", description = "Rule model")
public class RuleModel {
    @JsonPropertyDescription("Rule name")
    private String ruleName;

    @JsonPropertyDescription("Rule Python module content")
    private String rulePythonModuleContent;

    @JsonPropertyDescription("Rule runner type")
    private RuleRunnerType type = RuleRunnerType.python;

    @JsonPropertyDescription("Java class name for a rule runner that will execute the sensor. The \"type\" must be \"java_class\".")
    private String javaClassName = PythonRuleRunner.CLASS_NAME;

    @JsonPropertyDescription("Rule historic (past) values mode. A rule may require just the current sensor readout or use sensor readouts from past periods to perform prediction. " +
            "The number of time windows is configured in the time_window setting.")
    private RuleTimeWindowMode mode = RuleTimeWindowMode.current_value;

    @JsonPropertyDescription("Rule time window configuration when the mode is previous_readouts. Configures the number of past time windows (sensor readouts) that " +
            "are passes as a parameter to the rule. For example, to calculate the average or perform prediction on historic data.")
    private RuleTimeWindowSettingsSpec timeWindow;

    @JsonPropertyDescription("List of fields that are parameters of a custom rule. Those fields are used by the DQOps UI to display the data quality check " +
            "editing screens with proper UI controls for all required fields.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private ParameterDefinitionsListSpec fields;

    @JsonPropertyDescription("Additional rule parameters")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> parameters;

    @JsonPropertyDescription("This rule has a custom (user level) definition.")
    private boolean custom;

    @JsonPropertyDescription("This rule has is a built-in rule.")
    private boolean builtIn;

    /**
     * Boolean flag that decides if the current user can update or delete this object.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete this object.")
    private boolean canEdit;

    /**
     * Optional parsing error that was captured when parsing the YAML file.
     * This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.
     */
    @JsonPropertyDescription("Optional parsing error that was captured when parsing the YAML file. " +
            "This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing error message and the file location.")
    private String yamlParsingError;

    /**
     * Default constructor for RuleModel.
     */
    public RuleModel() {
    }

    /**
     * Constructor for RuleModel with a given {@link RuleDefinitionWrapper}, custom flag, and built-in flag.
     * Sets the properties of the rule based on the given {@link RuleDefinitionWrapper}.
     * @param ruleDefinitionWrapper The RuleDefinitionWrapper to use.
     * @param custom Whether the rule is a custom rule.
     * @param builtIn Whether the rule is a built-in rule.
     * @param canEdit The current user can edit the definition of the rule.
     */
    public RuleModel(RuleDefinitionWrapper ruleDefinitionWrapper, boolean custom, boolean builtIn, boolean canEdit) {
        this.ruleName = ruleDefinitionWrapper.getRuleName();
        this.rulePythonModuleContent = ruleDefinitionWrapper.getRulePythonModuleContent().getTextContent();
        RuleDefinitionSpec ruleDefinitionSpec = ruleDefinitionWrapper.getSpec();
        this.type = ruleDefinitionSpec.getType();
        this.javaClassName = ruleDefinitionSpec.getJavaClassName();
        this.mode = ruleDefinitionSpec.getMode();
        this.timeWindow = ruleDefinitionSpec.getTimeWindow();
        this.fields = ruleDefinitionSpec.getFields();
        this.parameters = ruleDefinitionSpec.getParameters();
        this.custom = custom;
        this.builtIn = builtIn;
        this.canEdit = canEdit;
        this.yamlParsingError = ruleDefinitionSpec.getYamlParsingError();
    }

    /**
     * Checks whether the RuleModel is equal to a given {@link RuleDefinitionWrapper}.
     * @param ruleDefinitionWrapper The RuleDefinitionWrapper to compare.
     * @return true if the RuleModel is equal to the RuleDefinitionWrapper, false otherwise.
     */
    public boolean equalsBuiltInRule(RuleDefinitionWrapper ruleDefinitionWrapper) {
        if (ruleDefinitionWrapper == null) {
            return false;
        }

        if (!Objects.equals(Optional.of(ruleDefinitionWrapper.getRulePythonModuleContent()).orElse(new FileContent()).getTextContent(),
                rulePythonModuleContent)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getType(), type)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getJavaClassName(), javaClassName)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getMode(), mode)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getParameters(), parameters)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getFields(), fields)) {
            return false;
        }

        if (!Objects.equals(ruleDefinitionWrapper.getSpec().getTimeWindow(), timeWindow)) {
            return false;
        }

        return true;
    }

    /**
     * Returns the rule definition spec object.
     * @return the rule definition spec object.
     */
    public RuleDefinitionSpec toRuleDefinitionSpec() {

        RuleDefinitionSpec ruleDefinitionSpec = new RuleDefinitionSpec();
        ruleDefinitionSpec.setType(type);
        ruleDefinitionSpec.setJavaClassName(javaClassName);
        ruleDefinitionSpec.setMode(mode);
        ruleDefinitionSpec.setTimeWindow(timeWindow);
        ruleDefinitionSpec.setFields(fields);
        ruleDefinitionSpec.setParameters(parameters);

        return ruleDefinitionSpec;
    }

    /**
     * Returns the file content object with Python module content.
     * @return file content object with Python module content.
     */
    public FileContent makePythonModuleFileContent() {
        return new FileContent(rulePythonModuleContent);
    }

    public static class RuleModelSampleFactory implements SampleValueFactory<RuleModel> {
        @Override
        public RuleModel createSample() {
            return new RuleModel() {{
                setRuleName(SampleStringsRegistry.getRuleName());
                setCanEdit(true);
            }};
        }
    }
}

File: distribution/python/dqops/client/models/rule_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.rule_runner_type import RuleRunnerType
from ..models.rule_time_window_mode import RuleTimeWindowMode
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.parameter_definition_spec import ParameterDefinitionSpec
    from ..models.rule_model_parameters import RuleModelParameters
    from ..models.rule_time_window_settings_spec import RuleTimeWindowSettingsSpec


T = TypeVar("T", bound="RuleModel")


@_attrs_define
class RuleModel:
    """Rule model

    Attributes:
        rule_name (Union[Unset, str]): Rule name
        rule_python_module_content (Union[Unset, str]): Rule Python module content
        type (Union[Unset, RuleRunnerType]):
        java_class_name (Union[Unset, str]): Java class name for a rule runner that will execute the sensor. The "type"
            must be "java_class".
        mode (Union[Unset, RuleTimeWindowMode]):
        time_window (Union[Unset, RuleTimeWindowSettingsSpec]):
        fields (Union[Unset, List['ParameterDefinitionSpec']]): List of fields that are parameters of a custom rule.
            Those fields are used by the DQOps UI to display the data quality check editing screens with proper UI controls
            for all required fields.
        parameters (Union[Unset, RuleModelParameters]): Additional rule parameters
        custom (Union[Unset, bool]): This rule has a custom (user level) definition.
        built_in (Union[Unset, bool]): This rule has is a built-in rule.
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can update or delete this object.
        yaml_parsing_error (Union[Unset, str]): Optional parsing error that was captured when parsing the YAML file.
            This field is null when the YAML file is valid. If an error was captured, this field returns the file parsing
            error message and the file location.
    """

    rule_name: Union[Unset, str] = UNSET
    rule_python_module_content: Union[Unset, str] = UNSET
    type: Union[Unset, RuleRunnerType] = UNSET
    java_class_name: Union[Unset, str] = UNSET
    mode: Union[Unset, RuleTimeWindowMode] = UNSET
    time_window: Union[Unset, "RuleTimeWindowSettingsSpec"] = UNSET
    fields: Union[Unset, List["ParameterDefinitionSpec"]] = UNSET
    parameters: Union[Unset, "RuleModelParameters"] = UNSET
    custom: Union[Unset, bool] = UNSET
    built_in: Union[Unset, bool] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    yaml_parsing_error: Union[Unset, str] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        rule_name = self.rule_name
        rule_python_module_content = self.rule_python_module_content
        type: Union[Unset, str] = UNSET
        if not isinstance(self.type, Unset):
            type = self.type.value

        java_class_name = self.java_class_name
        mode: Union[Unset, str] = UNSET
        if not isinstance(self.mode, Unset):
            mode = self.mode.value

        time_window: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.time_window, Unset):
            time_window = self.time_window.to_dict()

        fields: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.fields, Unset):
            fields = []
            for fields_item_data in self.fields:
                fields_item = fields_item_data.to_dict()

                fields.append(fields_item)

        parameters: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.parameters, Unset):
            parameters = self.parameters.to_dict()

        custom = self.custom
        built_in = self.built_in
        can_edit = self.can_edit
        yaml_parsing_error = self.yaml_parsing_error

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if rule_name is not UNSET:
            field_dict["rule_name"] = rule_name
        if rule_python_module_content is not UNSET:
            field_dict["rule_python_module_content"] = rule_python_module_content
        if type is not UNSET:
            field_dict["type"] = type
        if java_class_name is not UNSET:
            field_dict["java_class_name"] = java_class_name
        if mode is not UNSET:
            field_dict["mode"] = mode
        if time_window is not UNSET:
            field_dict["time_window"] = time_window
        if fields is not UNSET:
            field_dict["fields"] = fields
        if parameters is not UNSET:
            field_dict["parameters"] = parameters
        if custom is not UNSET:
            field_dict["custom"] = custom
        if built_in is not UNSET:
            field_dict["built_in"] = built_in
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit
        if yaml_parsing_error is not UNSET:
            field_dict["yaml_parsing_error"] = yaml_parsing_error

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.parameter_definition_spec import ParameterDefinitionSpec
        from ..models.rule_model_parameters import RuleModelParameters
        from ..models.rule_time_window_settings_spec import RuleTimeWindowSettingsSpec

        d = src_dict.copy()
        rule_name = d.pop("rule_name", UNSET)

        rule_python_module_content = d.pop("rule_python_module_content", UNSET)

        _type = d.pop("type", UNSET)
        type: Union[Unset, RuleRunnerType]
        if isinstance(_type, Unset):
            type = UNSET
        else:
            type = RuleRunnerType(_type)

        java_class_name = d.pop("java_class_name", UNSET)

        _mode = d.pop("mode", UNSET)
        mode: Union[Unset, RuleTimeWindowMode]
        if isinstance(_mode, Unset):
            mode = UNSET
        else:
            mode = RuleTimeWindowMode(_mode)

        _time_window = d.pop("time_window", UNSET)
        time_window: Union[Unset, RuleTimeWindowSettingsSpec]
        if isinstance(_time_window, Unset):
            time_window = UNSET
        else:
            time_window = RuleTimeWindowSettingsSpec.from_dict(_time_window)

        fields = []
        _fields = d.pop("fields", UNSET)
        for fields_item_data in _fields or []:
            fields_item = ParameterDefinitionSpec.from_dict(fields_item_data)

            fields.append(fields_item)

        _parameters = d.pop("parameters", UNSET)
        parameters: Union[Unset, RuleModelParameters]
        if isinstance(_parameters, Unset):
            parameters = UNSET
        else:
            parameters = RuleModelParameters.from_dict(_parameters)

        custom = d.pop("custom", UNSET)

        built_in = d.pop("built_in", UNSET)

        can_edit = d.pop("can_edit", UNSET)

        yaml_parsing_error = d.pop("yaml_parsing_error", UNSET)

        rule_model = cls(
            rule_name=rule_name,
            rule_python_module_content=rule_python_module_content,
            type=type,
            java_class_name=java_class_name,
            mode=mode,
            time_window=time_window,
            fields=fields,
            parameters=parameters,
            custom=custom,
            built_in=built_in,
            can_edit=can_edit,
            yaml_parsing_error=yaml_parsing_error,
        )

        rule_model.additional_properties = d
        return rule_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
